bindings/cpp/src/ethereum.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::{anyhow, Result};
+use defi_wallet_core_common::contract::ContractCall;
+use defi_wallet_core_common::contract::DynamicContract;
+use defi_wallet_core_common::EthAbiTokenBind;
+use defi_wallet_core_common::EthError;
+use ethers::abi::Detokenize;
+use ethers::abi::InvalidOutputType;
+use ethers::abi::Token;
+use ethers::prelude::*;
+use ethers::types::transaction::eip2718::TypedTransaction;
+use std::convert::TryFrom;
+
+pub struct EthContract {
+    abi_json: String,
+}
+
+pub struct EthDetokenizer {
+    json: String,
+}
+impl Detokenize for EthDetokenizer {
+    fn from_tokens(tokens: Vec<Token>) -> std::result::Result<Self, InvalidOutputType>
+    where
+        Self: Sized,
+    {
+        let json = serde_json::to_string(&tokens)
+            .map_err(|e| InvalidOutputType(format!("serde json error {:?}", e,)))?;
+        Ok(EthDetokenizer { json })
+    }
+}
+
+#[cxx::bridge(namespace = "org::defi_wallet_core")]
+#[allow(clippy::too_many_arguments)]
+mod ffi {
+
+    extern "Rust" {
+        type EthContract;
+
+        fn new_eth_contract(abi_json: String) -> Result<Box<EthContract>>;
+
+        fn encode(
+            &mut self,
+            rpcserver: &str,
+            contract_address: &str,
+            function_name: &str,
+            function_args: &str, // json
+        ) -> Result<Vec<u8>>;
+
+        fn call(
+            &mut self,
+            rpcserver: &str,
+            contract_address: &str,
+            function_name: &str,
+            function_args: &str, // json
+        ) -> Result<String>;
+
+    }
+} // end of ffi
+
+fn new_eth_contract(abi_json: String) -> Result<Box<EthContract>> {
```

Comment:
should this take `contract_address` and `rpcserver`, so that `Provider` and `DynamicContract` can be constructed just once?

Reply:
ok